aptos-move/framework/aptos-framework/sources/voting.move

```diff
@@ -326,6 +373,67 @@ module aptos_framework::voting {
         option::extract(&mut proposal.execution_content)
     }
 
+    /// Resolve a single-step or a multi-step proposal with given id.
+    /// Can only be done if there are at least as many votes as min required and
+    /// there are more yes votes than no. If either of these conditions is not met, this will revert.
+    ///
+    /// @param voting_forum_address The address of the forum where the proposals are stored.
+    /// @param proposal_id The proposal id.
+    /// @param next_execution_hash The next execution hash if the given proposal is multi-step.
+    public fun resolve_single_step_or_multi_step_proposal<ProposalType: store>(
+        voting_forum_address: address,
+        proposal_id: u64,
+        next_execution_hash: vector<u8>,
+    ) acquires VotingForum {
+        let proposal_state = get_proposal_state<ProposalType>(voting_forum_address, proposal_id);
+        assert!(proposal_state == PROPOSAL_STATE_SUCCEEDED, error::invalid_state(EPROPOSAL_CANNOT_BE_RESOLVED));
+
+        let voting_forum = borrow_global_mut<VotingForum<ProposalType>>(voting_forum_address);
+        let proposal = table::borrow_mut(&mut voting_forum.proposals, proposal_id);
+        assert!(!proposal.is_resolved, error::invalid_state(EPROPOSAL_ALREADY_RESOLVED));
+
+        // If this proposal is multi-step, we update the metadata map to indicate that this proposal is in execution.
+        let multi_step_in_execution_key = utf8(IS_MULTI_STEP_PROPOSAL_IN_EXECUTION_KEY);
+        if (simple_map::contains_key(&proposal.metadata, &multi_step_in_execution_key)) {
```

Comment:
This is interesting. There's some redundancy between IS_MULTI_STEP_PROPOSAL_IN_EXECUTION_KEY and IS_MULTI_STEP_PROPOSAL:

1. If IS_MULTI_STEP_PROPOSAL_IN_EXECUTION_KEY is not present in the metadata, the proposal is single-step.
2. If IS_MULTI_STEP_PROPOSAL_IN_EXECUTION_KEY is present but false, the proposal is multi-step and has not entered execution phase.
3. If IS_MULTI_STEP_PROPOSAL_IN_EXECUTION_KEY is present but true, the proposal is multi-step and has entered execution phase.

I wonder if we can combine these 2 flags?

Reply:
I think we can do a key `PROPOSAL_TYPE_AND_STATE` and setting it to 
- value `SINGLE_STEP` if the proposal is single-step;
- value `MULTI_STEP` if the proposal is multi-step and not in execution;
- value `MULTI_STEP_IN_EXECUTION` if the proposal is multi-step and in execution.
However this might look messier because we're introducing three new constants.

Reply:
That's true. I'm fine with having 2 like what we have currently as well. Can you add some documentation above the metadata field in the struct to list the special metadata attributes that are used by the voting flow? We'll have 3 in total
1. Last voting time
2. proposal type (single vs multi-step)
3. Whether the proposal is in execution state.

Speaking of which, make sure the MULTI_STEP_IN_EXECUTION field is clear too after all steps have been executed. It's not crucial to the on chain flow but can be cleaner for off chain to interpret if it's closed out properly when execution is fully complete.

Reply:
Good point! Just added comments for these variables. Also cleared `MULTI_STEP_IN_EXECUTION` upon successful resolution of the multi-step proposal.